DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,552 @@
+
+            preOrderWalk = (node) =>
+            {
+                if (node == null)
+                {
+                    return;
+                }
+
+                result.Add(node.Key);
+                preOrderWalk!(node.Left);
+                preOrderWalk(node.Right);
+            };
+
+            preOrderWalk(root);
+            return result;
+        }
+
+        /// <summary>
+        /// Gets all the elements in the tree in post-order order.
+        /// </summary>
+        /// <returns>Sequence of elements in post-order order.</returns>
+        public IEnumerable<TKey> GetKeysPostOrder()
+        {
+            var result = new List<TKey>();
+
+            Action<AATreeNode<TKey>?>? postOrderWalk = null;
+
+            postOrderWalk = (node) =>
+            {
+                if (node == null)
+                {
+                    return;
+                }
+
+                postOrderWalk!(node.Left);
+                postOrderWalk(node.Right);
+                result.Add(node.Key);
+            };
+
+            postOrderWalk(root);
+            return result;
+        }
+
+        /// <summary>
+        /// Checks if the tree is a valid AA Tree.
+        /// </summary>
+        /// <returns>true if tree is a valid AA Tree, false otherwise.</returns>
+        public bool Validate()
+        {
+            return ValidateTree(root);
+        }
+
+        /// <summary>
+        /// Recursive function to add an element to the tree.
+        /// </summary>
+        /// <param name="key">The element to add.</param>
+        /// <param name="node">The node to search for a empty spot.</param>
+        /// <returns>The node with the added element.</returns>
+        /// <exception cref="ArgumentException">Thrown if key is already in the tree.</exception>
+        private AATreeNode<TKey> Add(TKey key, AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return new AATreeNode<TKey>(key, 1);
+            }
+            else if (comparer.Compare(key, node.Key) < 0)
+            {
+                node.Left = Add(key, node.Left);
+            }
+            else if (comparer.Compare(key, node.Key) > 0)
+            {
+                node.Right = Add(key, node.Right);
+            }
+            else
+            {
+                throw new ArgumentException($"Key \"{key}\" already in tree!", nameof(key));
+            }
+
+            return Split(Skew(node)) !;
+        }
+
+        /// <summary>
+        /// Recursive function to remove an element from the tree.
+        /// </summary>
+        /// <param name="key">The element to remove.</param>
+        /// <param name="node">The node to search from.</param>
+        /// <returns>The node with the specified element removed.</returns>
+        private AATreeNode<TKey>? Remove(TKey key, AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return node;
```

Comment:
When is this executed? This is not a good idea to return null, should probably throw an exception of some kind